target/linux/ramips/image/mt7621.mk

```diff
@@ -235,6 +235,22 @@ define Device/k2p
 endef
 TARGET_DEVICES += k2p
 
+define Device/mir3p
+  DTS := MIR3P
+  BLOCKSIZE := 128k
+  PAGESIZE := 2048
+  KERNEL_SIZE:= 4096k
+  UBINIZE_OPTS := -E 5
+  IMAGE_SIZE := $(ralink_default_fw_size_32M)
+  DEVICE_TITLE := Xiaomi Mi Router 3 Pro
+  IMAGES += factory.bin
+  IMAGE/sysupgrade.bin := sysupgrade-tar | append-metadata
+  IMAGE/factory.bin := append-kernel | pad-to $$$$(BLOCKSIZE) | append-ubi | check-size $$$$(IMAGE_SIZE)
```

Comment:
Hm, there is a problem here. Unlike the squashfs + jffs combo we know and love/hate on NOR chip. The ubi partition can't be moved once it is in place on the device. With your current setup, a user that installed the image will run into trouble (bricked units) when the kernel grows beyond the initially allocated. 

I think you probably have to pad-to the full KERNEL_SIZE here after append-kernel and rework the partition layout in the dts so that the ubi partition starts at a fixed 4M after the kernel.

Reply:
Excellent point. I was hoping someone would give me advice on how to do this right.

I've changed it to pad-to $$(KERNEL_SIZE) and it seems like that should be enough (ie, won't need to change the .dts file) since my "firmware" partition is "denx,uimage" and that gets split at boot-time and ubi comes after KERNEL_SIZE.

Please tell me if I'm missing something.

Reply:
> won't need to change the .dts file) since my "firmware" partition is "denx,uimage" and that gets split at boot-time and ubi comes after KERNEL_SIZE.

From what I can tell, the mtdsplit code that is involved mtd_find_rootfs_from() will look for a UBI-Magic (4 bytes) in the blocks after the kernel. So depending on the kernel size you could look at 20 to 40 of block reads before the magic is found. This can slow down the boot, but If your nand is reasonably quick it could very well be that nobody notice it.

Reply:
OK i see. a "performance" issue not a "correctness" one. i'll leave the decision up to you. the kernel is <2mb and blocksize is 128k so that will be about 20 block reads before finding the magic. i was under the impression that i *needed* to use the new "denx,uimage" way of doing things. i can just as easily add a "ubi" partition.

what's the verdict?

Reply:
I thought of it a little longer, and I think it's clear what your verdict is, and I've updated the .dts file accordingly. I keep a "firmware" partition that spans both "kernel" and "ubi".

I don't know if I'm supposed to post a new patch file (with the edits you've suggested) or not...